File: src/main/java/com/slackow/func/parser/value/Type.java
package com.slackow.func.parser.value;

public abstract class Type extends Value {

    protected Type() {
        super(TypeType.INSTANCE);
    }

    public abstract String getName();

    @Override
    public String toString() {
        return "[" + getName() + "]";
    }

    private static class TypeType extends Type {
        private static final TypeType INSTANCE = new TypeType();

        @Override
        public String getName() {
            return "type";
        }
    }
}

File: src/main/java/com/slackow/func/parser/value/Type.kt
package com.slackow.func.parser.value

abstract class Type : Value(TypeType) {
    abstract val typeName: String

    override fun toString(): String {
        return "[$typeName]"
    }

    private object TypeType : Type() {
        override val typeName: String
            get() = "type"
    }
}